src/node_api.cc

```diff
@@ -3551,3 +3552,233 @@ napi_status napi_run_script(napi_env env,
+    return status;
+  }
+
+  uv_loop_t* loop = uv_default_loop();
+
+  status = napi_generic_failure;
+
+  if (func_type == napi_function) {
+    ts_fn = new napi_threadsafe_function__;
+    if (ts_fn != nullptr) {
+      if (uv_async_init(loop, &ts_fn->async, napi_threadsafe_function_cb) ==
+          0) {
+        if (uv_mutex_init(&ts_fn->mutex) == 0) {
+          status = napi_create_reference(env, func, 1, &ts_fn->ref);
+          if (status == napi_ok) {
+            if (max_queue_size == 0 || uv_cond_init(&ts_fn->cond) == 0) {
+              if (uv_idle_init(loop, &ts_fn->idle) == 0) {
+                ts_fn->call_js_cb = call_js_cb == nullptr ?
+                    napi_threadsafe_function_default_call_js : call_js_cb;
+                ts_fn->hint = hint;
+                ts_fn->env = env;
+                ts_fn->max_queue_size = max_queue_size;
+                ts_fn->idle_running = false;
+
+                *result = reinterpret_cast<napi_threadsafe_function>(ts_fn);
+                return napi_clear_last_error(env);
+              }
+              if (max_queue_size != 0) {
+                uv_cond_destroy(&ts_fn->cond);
+              }
+            }
+            status = napi_delete_reference(env, ts_fn->ref);
+
+            // Although we have successfully deleted the reference, we have
+            // failed to create the thread-safe function;
+            if (status == napi_ok) {
+              status = napi_generic_failure;
+            }
+          }
+          uv_mutex_destroy(&ts_fn->mutex);
+        }
+        uv_close(reinterpret_cast<uv_handle_t*>(&ts_fn->async),
+            [] (uv_handle_t* handle) -> void {
+              delete handle;
+            });
+
+        // Prevent the thread-safe function from being deleted here, because the
+        // uv_close() callback above will delete it.
+        ts_fn = nullptr;
+      }
+      delete ts_fn;
+    }
+  }
+  return napi_set_last_error(env, status);
+}
+
+NAPI_EXTERN napi_status
+napi_call_threadsafe_function(napi_threadsafe_function async,
+                              void* data,
+                              bool* result) {
+  bool local_result = false;
+  napi_status status = napi_ok;
+  uv_mutex_lock(&async->mutex);
+  while (async->queue.size() >= async->max_queue_size &&
+      async->max_queue_size > 0) {
+    if (result != nullptr) {
+      goto done;
+    }
+    uv_cond_wait(&async->cond, &async->mutex);
+  }
+  status = uv_async_send(&async->async) == 0 ? napi_ok : napi_generic_failure;
+  if (status == napi_ok) {
+    async->queue.push(data);
+    local_result = true;
+  }
+done:
+  if (result != nullptr) {
+    *result = local_result;
+  }
+  uv_mutex_unlock(&async->mutex);
+  return status;
+}
+
+NAPI_EXTERN napi_status
+napi_delete_threadsafe_function(napi_env env,
+                                napi_threadsafe_function ts_fn) {
+  CHECK_ENV(env);
+  CHECK_ARG(env, ts_fn);
+
+  napi_status status = napi_delete_reference(env, ts_fn->ref);
```

Comment:
It probably makes sense to wait for the queue to drain:
* I can protect myself from the race between delete and call, but I can't protect myself from the main event loop draining remaining item clashing with delete. I might be able to keep a count and track my own pending items, and only call delete when the count is 0 (with proper locks), but I can only protect my own function call and when the function returns all bets are off. IMHO this should be handled by the API layer instead of pushing it to all applications that consume this API. 
* The function calls in the queue should be eventually gets drained - otherwise user data can leak.

Reply:
This shouldn't result in a crash if the `napi_threadsafe_function` is used as documented - that is, the threads are terminated before `napi_delete_threadsafe_function()` is called. For that same reason I need not lock the mutex. `napi_delete_threadsafe_function()` must always be called from the loop thread, so it's synchronous wrt. the idle callback and async callback. So, after `napi_delete_threadsafe_function()` completes libuv guarantees that there will be no more calls to the idle handler which would access the queue nor to the async callback which would set up an idle handler.

The items in the queue would indeed leak in such a case.

I don't think it's necessarily a good idea to wait for the queue to drain. I think being able to abort is a good thing - perhaps if there's a fatal error somewhere. In normal usage, the user would be aware of the consequences of deleting the thread-safe function.

So, I think we should either document that it's the user's responsibility to avoid leaks perhaps by making the last item on the queue be a terminator, or provide a delete-once-empty API in addition to `napi_delete_threadsafe_function()`, like `napi_delete_threadsafe_function_deferred()`.